esmvalcore/_data_finder.py

```diff
@@ -108,23 +187,89 @@ def get_start_end_year(filename):
     return int(start_year), int(end_year)
 
 
-def select_files(filenames, start_year, end_year):
-    """Select files containing data between start_year and end_year.
-
-    This works for filenames matching *_YYYY*-YYYY*.* or *_YYYY*.*
+def _parse_period(timerange):
+    start_date = None
+    end_date = None
+    time_format = None
+    datetime_format = (
+        isodate.DATE_BAS_COMPLETE + 'T' + isodate.TIME_BAS_COMPLETE)
+    if timerange.split('/')[0].startswith('P'):
+        try:
+            end_date = isodate.parse_datetime(timerange.split('/')[1])
+            time_format = datetime_format
+        except isodate.ISO8601Error:
+            end_date = isodate.parse_date(timerange.split('/')[1])
+            time_format = isodate.DATE_BAS_COMPLETE
+        delta = isodate.parse_duration(timerange.split('/')[0])
+        start_date = end_date - delta
+    elif timerange.split('/')[1].startswith('P'):
+        try:
+            start_date = isodate.parse_datetime(timerange.split('/')[0])
+            time_format = datetime_format
+        except isodate.ISO8601Error:
+            start_date = isodate.parse_date(timerange.split('/')[0])
+            time_format = isodate.DATE_BAS_COMPLETE
+        delta = isodate.parse_duration(timerange.split('/')[1])
+        end_date = start_date + delta
+
+    if time_format == datetime_format:
+        start_date = isodate.datetime_isoformat(
+            start_date, format=datetime_format)
+        end_date = isodate.datetime_isoformat(
+            end_date, format=datetime_format)
+    elif time_format == isodate.DATE_BAS_COMPLETE:
+        start_date = isodate.date_isoformat(start_date, format=time_format)
+        end_date = isodate.date_isoformat(end_date, format=time_format)
+
+    return start_date, end_date
+
+
+def _compare_dates(date, file_date):
+    date = re.sub("[^0-9]", '', date)
+    file_date = re.sub("[^0-9]", '', file_date)
+    if len(date) < len(file_date):
+        file_date = file_date[0:len(date)]
+    elif len(date) > len(file_date):
+        date = date[0:len(file_date)]
+
+    return date, file_date
+
+
+def select_files(filenames, timerange):
+    """Select files containing data between a given timerange.
+
+    If the timerange is given as a period, the file selection
+    occurs taking only the years into account.
+
+    Otherwise, the file selection occurs taking into account
+    the time resolution of the file.
     """
     selection = []
+    start_date, end_date = _parse_period(timerange)
+
+    if start_date is None and end_date is None:
+        start_date = timerange.split('/')[0]
+        end_date = timerange.split('/')[1]
+    else:
+        start_date = str(start_date)
+        end_date = str(end_date)
```

Comment:
Why are these lines not part of the _parse_period function? Now this code is duplicated in several places where the _parse_period function is used.